addons/field_rations/functions/fnc_addRefillActions.sqf

```diff
@@ -0,0 +1,49 @@
+/*
+ * Author: mharis001, Glowbal, PabstMirror
+ * Adds refill water actions to water sources.
+ *
+ * Arguments:
+ * None
+ *
+ * Return Value:
+ * None
+ *
+ * Example:
+ * [] call acex_field_rations_fnc_addRefillActions
+ *
+ * Public: No
+ */
+#include "script_component.hpp"
+
+// Get config based classes for water sources to allow addition of more sources
+private _filter = QUOTE(getNumber (_x >> 'scope') == 2 && {getNumber (_x >> QQGVAR(refillSource)) > 0});
+private _waterSources = _filter configClasses (configFile >> "CfgVehicles");
+
+private _fnc_getChildren = {
+    params ["", "_player"];
+
+    private _actions = [];
+    {
+        private _cfg = configFile >> "CfgWeapons" >> _x;
+
+        // Every item that can be refilled, will have an action
+        if (isClass _cfg && {getText (_cfg >> QGVAR(onRefill)) != ""}) then {
+            private _displayName = getText (_cfg >> "displayName");
+            private _picture = getText (_cfg >> "picture");
+            private _actionText = format ["%1 %2", localize LSTRING(Refill), _displayName];
+            private _action = [_x, _actionText, _picture, LINKFUNC(refillItem), LINKFUNC(canRefillItem), {}, _x] call ACEFUNC(interact_menu,createAction);
+            _actions pushBack [_action, [], _player];
+        };
+    } forEach (items _player);
```

Comment:
This is EXTREMELY bad. `items player` might contain hundreds of duplicate items that are of no interest to you here.
I made that mistake in TFAR. Won't do it again. Playing with 10 fps and then suddenly having 50 when just dropping the backpack is not fun.
Best use would be the new `CBA_fnc_uniqueUnitItems`

Reply:
`CBA_fnc_uniqueUnitItems` adds `assignedItems` which isn't needed for this. `_items = items _player; _items arrayIntersect _items` to remove duplicates should fix this issue?

Reply:
I guess these actions really should be cached, with a `CBA_fnc_addPlayerEventHandler` `loadout` event clearing the cache.

Reply:
interaction menu updates only updates actions every second or so and only when the menu is open
so it's not that bad imho
Could cache the filtered list via something like
```
preInit:
GVAR(waterAllItems) = [-1];
GVAR(waterFilteredItems) = [];


loop:
private _loopOn = if (!(GVAR(waterAllItems) isEqualTo (items player))) then {
    GVAR(waterAllItems) = items player;
    GVAR(waterFilteredItems) = GVAR(waterAllItems) select {...};
};

{
...
} forEcah GVAR(waterFilteredItems);
```

Reply:
I was thinking about using `ace_common_fnc_cachedCall` with a duration of 5 seconds (with loadout event clearing it).

Reply:
`items` does not include what `assignedItems` returns, and...
> _items arrayIntersect _items is also relatively slow. Compared to just copying the uniqueItems code and removing the assignedItems part

the CBA uniqueItems function uses `arrayIntersect` too?

Reply:
yes. But CBA uniqueItems uses it on a couple dozen items. Not on hundreds.

Reply:
Here are absolute (with vanilla uniform/vest/backpack, modded can ofcause get bigger) worst case:
`items player` 0.266ms
`[player, false] call CBA_fnc_uniqueUnitItems` 0.124ms
```sqf
_allItems = items player;
_allItems arrayIntersect _allitems;
```
2.002ms
and just because you don't want assignedItems.

`([player, false] call CBA_fnc_uniqueUnitItems) - assignedItems player` 0.130ms

Reply:
And you have the same issue in https://github.com/acemod/ACEX/pull/114/files/7a5d0d1bddc7262eac1358e81efa0b27ca413ed6#diff-e7997fb46fe6c1fc254eac74a9ac58d9R11
Which is not like Pabst says
>and only when the menu is open

Reply:
I just noticed a typo in the uniqueUnitItems func. 
I'll add a way to disable the assignedItems.

Reply:
Took a closer look at `uniqueUnitItems`, I see what your saying. It doesn't have one entry for every individual item since it uses `getItemCargo`, making it much faster. Thanks for the detailed explanation.

Reply:
Didn't realize using `items` was that big of a performance difference too.